Implements K function estimators

mirca opened the pull request:
This is PR partially address #5139.

~~This PR implements a simple estimator of the Ripley's K function. The implemented estimator does not take into account edge effects for now, but I plan to add at least one method.~~

This PR implements some estimators for Ripley's K function. Different strategies for edge-effect correction are considered. 

The current behavior is:

```
Kest = RipleysKEstimator(area=area, ...)
r = np.linspace(0, 2.5, 100)
Kest(data=data, radii=r, mode='edge-effect-method')
```
~~I am going to add this keyword ``mode`` to separate different algorithms for edge effects correction.~~

The keyword ``mode`` accounts for the different algorithm for edge effects correction.

Anyone interested on giving feedback on the interface? :)

cc @skendrew @crawfordsm

skendrew commented:
Great! Yes I'm interested in testing this, thanks for tagging me. I have some code I'd previously written, which included one edge correction strategy. I'll try out the code with my own data and report back.

mirca commented:
@skendrew I've added the ``translation`` mode. Looks similar against R's implementation.

![figure_1](https://cloud.githubusercontent.com/assets/13077051/22106175/2f8f9c6c-de26-11e6-8e72-6a011fdf8eb9.png)
![screen shot 2017-01-19 at 9 04 40 am](https://cloud.githubusercontent.com/assets/13077051/22106206/5ce0e0e0-de26-11e6-8d7c-413261c293e9.png)

mirca commented:
@skendrew Some more results against R:
![screen shot 2017-01-19 at 7 03 27 pm](https://cloud.githubusercontent.com/assets/13077051/22127351/540d243e-de7a-11e6-95e0-b1037a1ba7ed.png)
![figure_1](https://cloud.githubusercontent.com/assets/13077051/22127698/db402798-de7b-11e6-943d-b20687dcd24d.png)

R code
```
X <- runifpoint(500)
K <- Kest(X, correction=c('none', 'translate'))
plot(K)
```

Python code
```
import numpy as np
import matplotlib.pyplot as plt
from astropy.stats import RipleysKEstimate

x = np.random.uniform(low=0, high=1, size=500)
y = np.random.uniform(low=0, high=1, size=500)
z = np.array([x, y]).T

Kest = RipleysKEstimate(data=z, area=1, max_height=1, max_width=1)

r = np.linspace(0, 0.25, 100)
plt.plot(r, Kest(r, mode='translation'), color='black', label=r'$K_{trans}$')
plt.plot(r, Kest(r, mode='none'), color='red', ls='--', label=r'$K_{un}$')
plt.plot(r, Kest.poisson(r), color='green', ls=':', label=r'$K_{pois}$')
plt.legend(loc=2)
```

mirca commented:
I think this is good for a more specialized review. I have implemented a few edge-correction methods available in the literature (some of which are also implemented in R ``spatstat``). Please, also let me know whether you like the current interface of the code :)

P.S.: still working on the narrative docs.

cc @skendrew @crawfordsm

MSeifert04 commented:
There are two sphinx warnings that directly relate to this PR:

```
/home/travis/build/astropy/astropy/docs/stats/ripleyskestimate.rst:19: WARNING: Duplicate explicit target name: "1".

checking consistency... /home/travis/build/astropy/astropy/docs/stats/ripleyskestimate.rst:: WARNING: document isn't included in any toctree
```

MSeifert04 commented:
I could offer some performance improvements, I don't know if they are needed though.

mirca commented:
@MSeifert04 Do point them out, please =D

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,314 @@
+
+    @y_max.setter
+    def y_max(self, value):
+        if value is None or (isinstance(value, (float, int)) and value > 0):
```
That's mostly a stylistic comment, but I generally prefer `if ... elif ... else` in such cases:

```
if value is None:
    self._y_max = None
elif isinstance(value, (float, int)) and value > 0:
    self._y_max = value
else:
    # err
```

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,314 @@
+        return self.evaluate(data=data, radii=radii, mode=mode)
+
+    def _pairwise_diffs(self, data):
+        npts = len(data)
```
The inner `for`-loop could be vectorized, if the sample is big this will be much faster:

```
npts = len(data)
diff2 = np.zeros(shape=(npts * (npts - 1) // 2, 2), 
                 dtype=np.float64)
k = 0
for i in range(npts - 1):
    size = npts-i-1
    diff2[k: k+size] = abs(data[i] - data[i+1:])
    k += size
```

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,314 @@
+        return diff
+
+    def _pairwise_distances(self, data):
+        npts = len(data)
```
This function could be reduced to two lines:

```
diff = self._pairwise_diffs(data)
return np.hypot(diff[:, 0], diff[:, 1])
```

MSeifert04 replied:
`scipy` contains some distance calculation routines, maybe one could resort to them if `scipy` is installed:

```
return scipy.spatial.distance.pdist(data)
```

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,314 @@
+            Ripley's K function evaluated at ``radii``.
+        """
+
+        return np.pi * radii * radii
```
I know this formula under the name of "area of a circle", when did they change that to "poisson"? 😄 Just joking.

mirca replied:
Hahahaha xD

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,314 @@
+        elif mode == 'translation':
+            diff = self._pairwise_diffs(data)
+            distances = self._pairwise_distances(data)
+            intersec_area = (((self.x_max - self.x_min) - diff[:][:, 0]) *
```
Why the extra `[:]` in `diff[:][:, 0]` and `diff[:][:, 1]`?

mirca replied:
oops, nice catch, thanks =)

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,314 @@
+        # eq. 15.11 Stoyan book page 283
+        elif mode == 'translation':
+            diff = self._pairwise_diffs(data)
+            distances = self._pairwise_distances(data)
```
That seems like an unnecessary recalculation because `self._pairwise_distances` already calculates the `diffs` so these are calculated two times. But I have currently no suggestion how to improve that. :)

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,314 @@
+                    (self.x_max - self.x_min) / (self.y_max - self.y_min))
+            x = distances / math.sqrt(a / b)
+            u = np.sqrt((x * x - 1) * (x > 1))
+            v = np.sqrt((x * x - b ** 2) * (x < math.sqrt(b**2 + 1)) * (x > b))
```
Until now I assumed you did `x * x` because it could be more efficient than `x**2` but in this line you do both `x * x - b ** 2`, maybe stick to one style? If you really want to improve the performance here you could calculate some calculations up front:

```
x_squared = x*x
b_squared = b*b
x_greater_than_b = x > b
```

because these are reused multiple times. But it shouldn't make to much of a difference here. :)

mirca replied:
``b`` is always scalar and ``x`` is an ``np.array`` that's why I used different styles ;)
But yeah, I'll definitely store some calculations like ``x_greater_than_b = x > b``, thanks! =)

mirca commented:
@MSeifert04 Thank you a lot! Working on your suggestions :)

bsipocz commented on docs/stats/ripley.rst:
```diff
@@ -0,0 +1,58 @@
+
+.. plot::
+    :include-source:
+    import numpy as np
```
you need to have an empty line between the directive and code

mirca replied:
nice catch! thanks @bsipocz

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,307 @@
+        elif mode == 'translation':
+            diff = self._pairwise_diffs(data)
+            distances = np.hypot(diff[:, 0], diff[:, 1])
+            intersec_area = (((self.x_max - self.x_min) - diff[:][:, 0]) *
```
The extra `[:]` are still there.

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,307 @@
+            diff = self._pairwise_diffs(data)
+            distances = np.hypot(diff[:, 0], diff[:, 1])
+            intersec_area = (((self.x_max - self.x_min) - diff[:][:, 0]) *
+                             ((self.y_max - self.y_min) - diff[:][:, 1]))
```
here as well.

MSeifert04 commented on astropy/stats/spatial.py:
```diff
@@ -0,0 +1,307 @@
+            for k in range(npts - 1):
+                min_hor_dist = min(self.x_max - data[k][0], data[k][0] - self.x_min)
+                min_ver_dist = min(self.y_max - data[k][1], data[k][1] - self.y_min)
+                hor_dist = np.append(hor_dist, min_hor_dist * np.ones(npts - 1 - k))
```
Instead of `append` you could also initialize the final-sized array before and just set the slices. The values should be known, right?

mirca replied:
Oh yeah, thank you! @MSeifert04

mirca commented:
@MSeifert04 Thank you, I am going to get back to this soon!

mirca commented:
@MSeifert04 done! Please take a look :)

mirca commented:
Anyone want to take a look? I think this is ready for a final review ;-)
cc @crawfordsm @skendrew @MSeifert04 @larrybradley

crawfordsm commented:
Thanks @mirca! I'll give it a final review tomorrow and take a look through everything.

skendrew commented:
I'm checking it through too and seems to work great so far! One comment is that I tried to enter negative spatial coordinates and then got an error that x_min can't be negative. It would be great if that was possible, as users will often want to pass astronomical coordinates.

Traceback (most recent call last):
  File "ripleysk_test.py", line 27, in <module>
    plt.figure(figsize=[6,6])
  File "/Users/kendrew/milkywayproj/ripleys_k/astropy/astropy/stats/spatial.py", line 59, in __init__
    self.x_min = x_min
  File "/Users/kendrew/milkywayproj/ripleys_k/astropy/astropy/stats/spatial.py", line 120, in x_min
    'Got {}.'.format(value))
ValueError: x_min is expected to be a nonnegative number. Got -10.0.

mirca commented:
@crawfordsm @skendrew can you give another try on this? ;-)

bsipocz commented:
@crawfordsm - Could you please do a final review here?

mirca commented:
Anyone able to do a final review on this on time for 2.0? ;-)

cc @bsipocz @crawfordsm @MSeifert04 @skendrew

crawfordsm commented:
@skendrew will provide a final check this week, otherwise, this is ready to be merged.

mirca commented:
@skendrew @crawfordsm @eteq thanks ;)

crawfordsm commented:
Thanks @mirca !

mirca commented:
@crawfordsm Thanks Steve!

MSeifert04 commented:
Thanks @mirca! 🎉

mirca commented:
Thanks @MSeifert04!!
